### ISSUE TITLE
[feature] MQTT Proxy for Message Queue

### ISSUE THREAD

**Issue by suchen-sci** [megaease/easegress#187]
# Background
- MQTT is a standard messaging protocol for IoT (Internet of Things) which is extremely lightweight and used by a wide variety of industries.
- To support MQTT Proxy in Easegress, MQTT clients can produce messages to Kafka backend directly.

# Proposal
- Bring `MQTTProxy` as an BusinessController to Easegress
- Use `github.com/eclipse/paho.mqtt.golang/packets` to parse MQTT packet. `paho.mqtt.golang` is a MQTT 3.1.1 go client introduced by Eclipse Foundation (who also introduced the most widely used MQTT broker mosquitto).

# Design
1. Spec
```yaml
kind: MQTTProxy
name: mqttproxy
port: 1883  # tcp port for mqtt clients to connect 
backendType: Kafka
kafkaBroker:
  backend: ["123.123.123.123:9092", "234.234.234.234:9092"]
useTLS: true
certificate:
  - name: cert1
    cert: balabala
    key: keyForbalabala
  - name: cert2
    cert: foo
    key: bar
auth:
  # username and password for mqtt clients to connect broker 
  - userName: test
    passBase64: dGVzdA==
  - userName: admin
    passBase64: YWRtaW4=
topicMapper:
  # detail described in following 
  matchIndex: 0
  route: 
    - name: gateway
      matchExpr: "gate*"
    - name: direct
      matchExpr: "dir*"
  policies:
    - name: direct
      topicIndex: 1
      route: 
      - topic: iot_phone
        exprs: ["iphone", "xiaomi", "oppo", "pixel"]
      - topic: iot_other
        exprs: [".*"]  
      headers: 
      0: direct
      1: device
      2: status
    - name: gateway
      topicIndex: 3
      route: 
      - topic: iot_phone
        exprs: ["iphone", "xiaomi", "oppo", "pixel"]
      - topic: iot_other
        exprs: [".*"]  
      headers: 
      0: gateway
      1: gatewayID
      2: device
      3: status
```
More detail about topic mapping, in MQTT, there are multi-levels in topic, for example, `direct/iphone/log`. Topic mapping is used to map MQTT topic to Kafka topic with headers. 

Consider there may be multiple schemas for your MQTT topic, so we first provide a router to route your MQTT topic to different mapping policies and then do the map in that policy.

For example,
```
...
    matchIndex: 0
    route: 
    - name: gateway
        matchExpr: "gate*"
    - name: direct
        matchExpr: "dir*"
...
```
means that we use MQTT topic level 0 to match `matchExpr` to find corresponding policy. In this case, `gateway/gate123/iphone/log` will match policy `gateway`, `direct/iphone/log` will match policy `direct`. 

More example about topic mapper: 
```
use yaml above:
MQTT topic pattern1: gateway/gatewayID/device/status -> match policy gateway
MQTT topic pattern2: direct/device/status -> match policy direct

example1: "gateway/gate123/iphone/log"
Kafka 
    topic: iot_phone
    headers: 
        gateway: gateway
        gatewayID: gate123
        device: iphone
        status: log

example2: "direct/xiaomi/status"
Kafka 
    topic: iot_phone
    headers: 
        direct: direct
        device: xiaomi
        status: status

example3: "direct/tv/log"
Kafka 
    topic: iot_other
    headers: 
        direct: direct
        device: tv
        status: log
```
Empty `topicMapper` means there are no map between MQTT topic and Kafka topic.

2. Request sequence
```
 Client                Server               Kafka
+--------+            +--------+           +-------+
|        |1. connect  |        |           |       |     
|        |----------->|        |           |       |     
|        |            | check  |           |       |     
|        |2. connack  | info   |           |       |     
|        |<-----------|        |           |       |     
|        |            |        |           |       |  
|        |            |        |           |       |   
|        |3. publish  |        |4. send to |       |     
|        |----------->| topic  |   kafka   |       |     
|        |            |transfer|---------->|       |     
|        |3.1 puback  |        |           |       |     
|        |if qos=1    |        |           |       | 
+--------+            +--------+           +-------+
```
other request like pingreq/pingrsp, subscribe/unsubscribe is quite similar.

Now, we make any MQTT/3.1.1 client able to connect to our server and publish message to kafka backend and we also achieve some related control logic like keep alive, client takeover etc.

We also add session module, which allow us to keep MQTT client subscribe information and publish message to MQTT client. The message publish to MQTT client will come from our HTTP server.

# HTTP endpoint
We now support backend to send messages back to MQTT clients through http endpoint.

API for http endpoint:
- Host: Easegress IP, for example `http://127.0.0.1`
- Port: Easegress API address, by default, `:2381`
- Path: `apis/v1/mqttproxy/{name}/topics/publish`, where name is the name of MQTT proxy
- Method: POST
- Body: 
```json
{
  "topic": "yourTopicName", 
  "qos": 1, // currently only support 0 and 1
  "payload": "dataPayload",
  "base64": false
}
```
To send binary data, you can encode your binary data base64 and send `base64` flag to `true`. Your client will receive the original binary data, we will do the decode. 
- Status code:
  - 200: Success
  - 400: StatusBadRequest, may wrong http method, or wrong data (qos send to illegal number) etc. 

The HTTP endpoint schema also work for multi-node deployment. Say you have 3 Easegress instances called `eg-0`, `eg-1`, `eg-2`, and your MQTT client connects to `eg-0`, if you send messages to `eg-1`, your client will receive the message too.

We also support wildcard subscription. 
For example, 
```
POST http://127.0.0.1:2381/apis/v1/mqttproxy/mqttproxy/topics/publish
{
  "topic": "Beijing/Phone/Update", 
  "qos": 1, // currently only support 0 and 1
  "payload": "time to update",
  "base64": false
}

the clients subscribe following topics will receive the message:
"Beijing/Phone/Update"
"Beijing/+/Update"
"Beijing/Phone/+"
"+/Phone/Update"
"Beijing/+/+"
"Beijing/#"
"+/+/+"
```

# References 
1. https://github.com/eclipse/paho.mqtt.golang
2. http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html

**Comment by benja-wu** [comment]
``` yaml
kind: MQTTProxy
name: mqttproxy
port: 1883 # 1883 for default, 8883 for encrypted, tcp only
backend: localhost:9092 # kafka backend
```

* Should we consider a backend type for indicating it's Kafka, in case we want to add another type of backend in the future?
* Propose to adding `backendType` and related fields. For Kafka, it may look like
```
backendType: Kafka 
kafkaBroker: localhost:9092       # this field will only be used in `Kafka` type backend. 
```

**Comment by suchen-sci** [comment]
Good idea, general backend interface will also help for testing!

**Comment by benja-wu** [comment]
And I notice in your design, we choose to use `github.com/eclipse/paho.mqtt.golang/packets`.
It might be better if we have a basic comparison between this and other open-source solutions or libs. :-)

**Comment by suchen-sci** [comment]
Great idea! Ben!

There are four main reason we choose `paho.mqtt.golang` over other open-source solutions.
1. `paho.mqtt.golang/packets` is powerful enough, we can easily convert mqtt tcp packets to mqtt structs and vice versa
2. The code in `packets` is clean and well decoupling , only serialization and deserialization tools included. Some other libs may put serialization tools with control logics.
3. Besides `paho.mqtt.golang`, Eclipse Foundation also provide client for C, Python, Java, embedded-c, Rust... The client used to send message to us is very likely use the tool provided by Eclipse Foundation. And we can easily compatible with them and keep up with future updates.
4. There are also some open-source libs use `paho.mqtt.golang/packets` to do the parse, for example, `https://github.com/fhmq/hmq`, in this case, use these libs to parse is same to use `paho`.

**Comment by suchen-sci** [comment]
Update Spec to allow MQTT auth and add mapping rules to map MQTT topic to Kafka Topic

**Comment by codering** [comment]
Support wildcards to subscribe to multiple topics ?

**Comment by suchen-sci** [comment]
> Support wildcards to subscribe to multiple topics ?

Hi there, the answer for your question is not yet. I am sorry that we currently not support wildcards to subscribe to multi topics.

The whole MQTT feature is still working in process now. Currently, we have finished the MQTT proxy part (MQTT client can connect and publish message to Easegress, and Easegress will do topic mapping, and put the message to backend Kafka).

In a couple of weeks, we will plan to support to send message back to MQTT client through HTTP endpoint. At that time, we will truly support MQTT client to subscribe topics, now it is dummy.
So, after that, we may consider to add more advanced MQTT features to Easegress, for example, wildcards for topic subscription. 

To sum up, our implementation is not quite a traditional MQTT broker or MQTT proxy, it's more like combination of these two. MQTT client from IoT devices send message to Easegress, Easegress processes these messages and sends them to backend through message queue like Kafka. The backend process these messages and send replies to Easegress through HTTP endpoint, and Easegress sends these replies to MQTT client.

I hope this can help with your question.
